test/fennec/udp_test.exs

```diff
@@ -33,18 +34,133 @@ defmodule Fennec.UDPTest do
+
+      resp = udp_communicate(udp, 0, req)
+
+      params = Format.decode!(resp)
+      assert %Params{class: :failure,
+                     method: :allocate,
+                     identifier: ^id,
+                     attributes: [error]} = params
+
+      assert %ErrorCode{code: 400} = error
+    end
+
+    test "fails with unknown attribute", ctx do
+      udp = ctx.udp
+      id = :crypto.strong_rand_bytes(12)
+      req = allocate_request(id, [
+        %RequestedTransport{protocol: :udp},
+        %Lifetime{duration: 5}
+      ])
+
+      resp = udp_communicate(udp, 0, req)
+
+      params = Format.decode!(resp)
+      assert %Params{class: :failure,
+                     method: :allocate,
+                     identifier: ^id,
+                     attributes: [error]} = params
+
+      assert %ErrorCode{code: 420} = error
+    end
+
+    test "returns response with IPv6 XOR relayed address attribute", ctx do
+      udp = ctx.udp
+      %{server_address: server_address, client_address: client_address} = udp
+      client_port = client_port(udp, 0)
+      id = :crypto.strong_rand_bytes(12)
+      req = allocate_request(id)
+
+      resp = udp_communicate(udp, 0, req)
+
+      params = Format.decode!(resp)
+      assert %Params{class: :success,
+                     method: :allocate,
+                     identifier: ^id,
+                     attributes: attrs} = params
+      [lifetime, mapped, relayed] = Enum.sort(attrs)
+
+      assert %Lifetime{duration: 600} = lifetime
+
+      assert %XORMappedAddress{address: ^client_address,
+                               port: ^client_port,
+                               family: :ipv6} = mapped
+
+      assert %XORRelayedAddress{address: ^server_address,
+                                port: relayed_port,
+                                family: :ipv6} = relayed
+      assert relayed_port != udp.server_port
+    end
+
+    test "returns error after second allocation", ctx do
+      udp = ctx.udp
+      id1 = :crypto.strong_rand_bytes(12)
+      id2 = :crypto.strong_rand_bytes(12)
+      req1 = allocate_request(id1)
+      req2 = allocate_request(id2)
+
+      _resp = udp_communicate(udp, 0, req1)
+      resp = udp_communicate(udp, 0, req2)
+
+      params = Format.decode!(resp)
+      assert %Params{class: :failure,
+                     method: :allocate,
+                     identifier: ^id2,
+                     attributes: [error]} = params
+      assert %ErrorCode{code: 437} = error
+    end
+
+    test "returns success after redundant allocation", ctx do
+      udp = ctx.udp
+      id = :crypto.strong_rand_bytes(12)
+      req = allocate_request(id)
+
+      _resp = udp_communicate(udp, 0, req)
+      resp = udp_communicate(udp, 0, req)
+
+      params = Format.decode!(resp)
+      assert %Params{class: :success,
+                     method: :allocate,
+                     identifier: ^id,
+                     attributes: attrs} = params
+      assert 3 = length(attrs)
```

Comment:
This is a correct number if attributes of a success response. Those attributes were checked in previous test case, so here I just check the message just "feels" like correct one, because it's not about this message.